### PR TITLE
RPC: Add maxfeerate and maxburnamount args to submitpackage

### LINKED ISSUE #28949: submitpackage: add args "maxfeerate" and "maxburnamount"
### Please describe the feature you'd like to see added.

Currently `submitpackage` does no sanity checks on submitted transactions.

These checks have been helpful more than a handful of times in the wild, and these arguments
are actively used by downstream wallets. We should figure out how to support them

### Is your feature related to a problem, if so please describe it.

_No response_

### Describe the solution you'd like

`maxfeerate`: this one is a bit tricky. I think the easiest way would be have a utility function that pulls in utxos to create the sigops-adjusted vsizes and individual fees? Then the entire package could be rejected if any exceed the set amount.

`maxburnamount`: This can be added on without any chain-context by scanning the outputs of each transaction and rejecting the package if any exceed the value.

### Describe any alternatives you've considered

_No response_

### Please leave any additional context

_No response_

### PR DESCRIPTION
**Pull request by instagibbs**
Resolves https://github.com/bitcoin/bitcoin/issues/28949

I couldn't manage to do it very cleanly outside of (sub)package evaluation itself, since it would change the current interface very heavily. Instead I threaded through the max fee argument and used that directly via ATMPArgs. From that perspective, this is somewhat a reversion from https://github.com/bitcoin/bitcoin/pull/19339. In a post-cluster mempool world, these checks could be consolidated to right after the given (ancestor) package is linearized/chunked, by just checking the feerate of the top chunk and rejecting the submission entirely if the top chunk is too high.

The implication here is that subpackages can be submitted to the mempool prior to hitting this new fee-based error condition.

### CODE REVIEW

**Comment by instagibbs** [comment]
rebased on https://github.com/bitcoin/bitcoin/pull/28848 and ready for review

**Comment by ismaelsadeeq** [comment]
Concept ACK

**Review by glozow — COMMENTED** [review]
Concept ACK

My ideal approach would be to not check client-side stuff inside validation, e.g. by adding a helper that loads coins (probably not extensible to chunk feerate but could maybe be a sufficient sanity check), or a test package accept + real submit (like the way `BroadcastTransaction` does it). But if it's unavoidable, then this seems fine. AFAICT it's extensible to checking chunk feerate in the future.

**Comment by kashifs** [comment]
Concept ACK

I compiled from source, read through each modified line of code, ran the automated tests, and modified `test_maxfeerate_maxburn_submitpackage` test to check `maxfeerate` and `maxburnamount` seperately

File: src/validation.cpp [hunk 1]
```diff
@@ -1295,6 +1312,15 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::
             results.emplace(ws.m_ptx->GetWitnessHash(), MempoolAcceptResult::Failure(ws.m_state));
             return PackageMempoolAcceptResult(package_state, std::move(results));
         }
+
+        // Individual modified feerate exceeded caller-deemed "sane" max; abort
```
Comment (ismaelsadeeq) [inline src/validation.cpp:1316]: My understanding of package submission is that a child with unconfirmed parents is bundled together as a package for incentive compatibility. Should this instead check the package fee rate against `maxfeerate`?

If I understand correctly, currently, we are checking the package transaction fee rate individually against `maxfeerate`. The child transactions might be greater than the `maxfeerate`, but as a package might not be. In this case, we might reject the child transactions and accept the subpackage, which might not be incentive compatible at the current fee estimates.

For example, suppose we want to fee bump tx A, with the arrangement below. `maxfeerate` is set to 50 s/vb.

```md
               All txs are 10vb
               first value is absolute fee
               second value is ancestor score
               ┌───────────┐
               │  A (10s)  │
               │ 1 s/vb    │
               └───────────┘
                     │
                     ▼
               ┌───────────┐
               │  B (90s)  │
               │ 5 s/vb    │
               └───────────┘
                     │
                     ▼
               ┌────────────┐
               │  C (1000s) │
               │ 36.6 s/vb  │
               └────────────┘

```
If `maxfeerate` is 50s/vb we will accept A-B and reject C because it's ancestor score is 100s/vb.
But if we are evaluating maxfeerate against package fee rate we will accept the whole package A-B-C

File: src/rpc/client.cpp [hunk 2]
```diff
@@ -127,6 +127,8 @@ static const CRPCConvertParam vRPCConvertParams[] =
     { "testmempoolaccept", 0, "rawtxs" },
     { "testmempoolaccept", 1, "maxfeerate" },
     { "submitpackage", 0, "package" },
+    { "submitpackage", 1, "maxfeerate" },
+    { "submitpackage", 2, "maxburnamount" },
```
Comment (ismaelsadeeq) [inline src/rpc/client.cpp:131]: Should enable passing `maxfeerate` and `maxburnamount` as a config variables since it is used by various RPC, downstream might prefer this?

File: src/rpc/mempool.cpp [hunk 3]
```diff
@@ -867,6 +875,19 @@ static RPCHelpMan submitpackage()
...
+                                                     CFeeRate(AmountFromValue(request.params[1]));
+            std::optional<CFeeRate> max_sane_feerate{max_raw_tx_fee_rate};
+            // 0-value is special; it's mapped to no sanity check
+            if (max_raw_tx_fee_rate == CFeeRate(0)) {
+                max_sane_feerate = std::nullopt;
+            }
+
+            // Burn sanity check is run with no context
+            const CAmount max_burn_amount = request.params[2].isNull() ? 0 : AmountFromValue(request.params[2]);
+
```
Comment (ismaelsadeeq) [inline src/rpc/mempool.cpp:890]: We did not document this anywhere but the default behavior  if `maxburnamount` is not passed it we are rejecting the package if there is any transactions with unspendable output greater than 0.
Maybe indicate it in the `RPCHelpMan` of `maxburnamount`?

---

Its also not indicated for `sendrawtransaction`.

File: test/functional/rpc_packages.py [hunk 4]
```diff
@@ -356,5 +357,33 @@ def test_submitpackage(self):
...
+
+        self.log.info("Submitpackage maxburnamount arg testing")
+        tx = tx_from_hex(chain_hex[1])
+        tx.vout[-1].scriptPubKey = b'a' * 10001 # scriptPubKey bigger than 10k IsUnspendable
+        chain_hex = [chain_hex[0], tx.serialize().hex()]
+        # burn test is run before any package evaluation; nothing makes it in and we get broader exception
+        assert_raises_rpc_error(-25, "Unspendable output exceeds maximum configured by user", node.submitpackage, chain_hex, 0, 0.1)
+        assert_equal(node.getrawmempool(), [])
+
+        # Turn off restrictions for both and send it; parent gets through as own subpackage
```
Comment (ismaelsadeeq) [inline test/functional/rpc_packages.py:382]: To turn off for maxfeerate you should pass 0 fee rate, you cant turn off the restriction of `maxburnamount`
nit: maybe reword to relax?
```suggestion
# Relax the restrictions and send it; parent gets through as own subpackage
```

File: src/rpc/mempool.cpp [hunk 5]
```diff
@@ -183,7 +183,7 @@ static RPCHelpMan testmempoolaccept()
             Chainstate& chainstate = chainman.ActiveChainstate();
             const PackageMempoolAcceptResult package_result = [&] {
                 LOCK(::cs_main);
-                if (txns.size() > 1) return ProcessNewPackage(chainstate, mempool, txns, /*test_accept=*/true);
+                if (txns.size() > 1) return ProcessNewPackage(chainstate, mempool, txns, /*test_accept=*/true, /*max_sane_feerate=*/{});
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply src/rpc/mempool.cpp:186]: `testmempoolaccept` is basically shoving all transactions, whether or not they're a pacakge, into `AcceptMultipleTransactions`.

I'm not sure it makes sense due to this.

File: src/validation.cpp (hunk 1 above)
Comment (instagibbs) [reply src/validation.cpp:1316]: Ignoring topo restrictions in `AcceptPackage` since `C` should have `A` as a direct ancestor, 

> The child transactions might be greater than the maxfeerate, but as a package might not be. In this case, we might reject the child transactions and accept the subpackage, which might not be incentive compatible at the current fee estimates.

This is prior to any relay, of course restricting feerates is incentive compatible once published on the open network!

> If maxfeerate is 50s/vb we will accept A-B and reject C because it's ancestor score is 100s/vb.
But if we are evaluating maxfeerate against package fee rate we will accept the whole package A-B-C

Assuming you meant C's *feerate* is 100 sat/vbyte: If mempool minfee is 5 sat/vbyte, A+B would be accepted(as they'd be evaluated in their own subpackage already), then C evaluated on its own `AcceptSingleTransaction`, and rejected.

In a post-cluster mempool world, this would likely be reduced to evaluating chunks only, and so you'd look at A+B+C only, and reject the whole thing.

File: src/rpc/client.cpp (hunk 2 above)
Comment (instagibbs) [reply src/rpc/client.cpp:131]: startup config? Maybe? Could open an issue

File: src/rpc/mempool.cpp (hunk 3 above)
Comment (instagibbs) [reply src/rpc/mempool.cpp:890]: added a mention of the default being 0

File: test/functional/rpc_packages.py (hunk 4 above)
Comment (instagibbs) [reply test/functional/rpc_packages.py]: taken

File: src/validation.cpp (hunk 1 above)
Comment (ismaelsadeeq) [reply src/validation.cpp:1316]: >Assuming you meant C's feerate is 100 sat/vbyte

Yes C fee rated is 100s/vb, edited thanks.

>This is prior to any relay, of course restricting feerates is incentive compatible once published on the open network

Do you mean incentive incompatible here? because in the given example if C is rejected A will be  added to the mempool with mining score of 5s/vb which is lower than whats intended.

File: src/rpc/client.cpp (hunk 2 above)
Comment (ismaelsadeeq) [reply src/rpc/client.cpp:131]: Done in https://github.com/bitcoin/bitcoin/issues/29217

File: src/validation.cpp (hunk 1 above)
Comment (instagibbs) [reply src/validation.cpp:1316]: > Do you mean incentive incompatible here? because in the given example if C is rejected (A,B) will be added to the mempool with mining score of 5s/vb which is lower than whats intended.

Comment
err right
 
src/rpc/client.cpp
Comment on lines +130 to +131
    { "submitpackage", 1, "maxfeerate" },
    { "submitpackage", 2, "maxburnamount" },
Member
@ismaelsadeeq ismaelsadeeq 2 days ago • 
Should we just enable passing maxfeerate and maxburnamount as a config variables since it is used three RPC's now sendrawtransaction, testmempoolaccept and submitpackage, downstream might prefer this?

Member
Author
@instagibbs instagibbs 2 days ago
startup config? Maybe? Could open an issue

Member
@ismaelsadeeq ismaelsadeeq 3 hours ago
Done in #29217

@instagibbs	Reply...
src/rpc/mempool.cpp
Comment on lines +889 to +890
            const CAmount max_burn_amount = request.params[2].isNull() ? 0 : AmountFromValue(request.params[2]);

Member
@ismaelsadeeq ismaelsadeeq 2 days ago • 
We did not document this anywhere but the default behavior if maxburnamount is not passed we are rejecting the package if there is any transactions with unspendable output greater than 0.
Maybe indicate it in the RPCHelpMan of maxburnamount?

Its also not indicated for sendrawtransaction.

Member
Author
@instagibbs instagibbs 2 days ago
added a mention of the default being 0

@instagibbs	Reply...
test/functional/rpc_packages.py
Outdated
        assert_raises_rpc_error(-25, "Unspendable output exceeds maximum configured by user", node.submitpackage, chain_hex, 0, 0.1)
        assert_equal(node.getrawmempool(), [])

        # Turn off restrictions for both and send it; parent gets through as own subpackage
Member
@ismaelsadeeq ismaelsadeeq 2 days ago
To turn off for maxfeerate you should pass 0 fee rate, you cant turn off the restriction of maxburnamount
nit: maybe reword to relax?

Suggested change
        # Turn off restrictions for both and send it; parent gets through as own subpackage
# Relax the restrictions and send it; parent gets through as own subpackage
Member
Author
@instagibbs instagibbs 2 days ago
taken

@instagibbs	Reply...
@instagibbs
RPC: Add maxfeerate and maxburnamount args to submitpackage 
bf6fbea
@instagibbs instagibbs force-pushed the 2023-11-submitpackage-max-fee-burn branch from e650eed to bf6fbea 
2 days ago
@ismaelsadeeq ismaelsadeeq mentioned this pull request 3 hours ago
Enable maxfeerate and maxburnamount as startup config options. #29217
Open
Merge state
Add more commits by pushing to the 2023-11-submitpackage-max-fee-burn branch on instagibbs/bitcoin.

All checks have passed
1 skipped and 15 successful checks
@github-actions
CI / test each commit (pull_request) Skipped
Details
@github-actions
CI / macOS 13 native, x86_64, no depends, sqlite only, gui (pull_request) Successful in 34m
Details
@github-actions
CI / Win64 native, VS 2022 (pull_request) Successful in 25m
Details
@cirrus-ci
32-bit CentOS, dash, gui Successful in 21m — Task Summary
Details
@cirrus-ci
ARM, unit tests, no functional tests Successful in 27m — Task Summary
Details
@cirrus-ci
ASan + LSan + UBSan + integer, no depends, USDT Successful in 40m — Task Summary
Details
@cirrus-ci
MSan, depends Successful in 29m — Task Summary
Details
@cirrus-ci
TSan, depends, gui Successful in 32m — Task Summary
Details
@cirrus-ci
Win64, unit tests, no gui tests, no functional tests Successful in 15m — Task Summary
Details
@cirrus-ci
fuzzer,address,undefined,integer, no depends Successful in 42m — Task Summary
Details
@cirrus-ci
lint Successful in 1m — Task Summary
Details
@cirrus-ci
macOS-cross 11.0, gui, no tests Successful in 20m — Task Summary
Details
@cirrus-ci
multiprocess, i686, DEBUG Successful in 36m — Task Summary
Details
@cirrus-ci
no wallet, libbitcoinkernel Successful in 14m — Task Summary
Details
@cirrus-ci
previous releases, qt5 dev package and depends packages, DEBUG Successful in 25m — Task Summary
Details
@cirrus-ci
tidy Successful in 46m — Task Summary
Details
Merging is blocked
The base branch restricts merging to authorized users. Learn more about protected branches.
 You’re not authorized to merge this pull request.
@instagibbs


Add a comment
Comment
 
Add your comment here...
 
Remember, contributions to this repository should follow its contributing guidelines and security policy.
 ProTip! Add .patch or .diff to the end of URLs for Git’s plaintext views.
Reviewers
@glozow
glozow
@ismaelsadeeq
ismaelsadeeq
Still in progress?
Assignees
No one assigned
Labels
RPC/REST/ZMQ
Projects
None yet
Milestone
No milestone
Development
Successfully merging this pull request may close these issues.

 submitpackage: add args "maxfeerate" and "maxburnamount"
5 participants
@instagibbs
@DrahtBot
@[ismaelsadeeq](https://github.com/ismaelsadeeq)
@kashifs
@[glozow](https://github.com/glozow)
Allow edits and access to secrets by maintainers

Comment (instagibbs) [reply src/validation.cpp:1316]: > Do you mean incentive incompatible here? because in the given example if C is rejected (A,B) will be added to the mempool with mining score of 5s/vb which is lower than whats intended.

correct, typo

**Comment by ajtowns** [comment]
Isn't this check a little bit later than ideal? Wouldn't it make more sense to check both maxfeerate and maxburnamount as part of `signrawtransactionwith*` and `walletprocesspsbt` so that you never create a tx that can lose your money, rather than creating it and hoping that it won't find its way onto the network via some method other than rpc with default args?

**Comment by glozow** [comment]
> Isn't this check a little bit later than ideal? Wouldn't it make more sense to check both maxfeerate and maxburnamount as part of `signrawtransactionwith*` and `walletprocesspsbt` so that you never create a tx that can lose your money

I think it's indeed good to have that, but isn't it still possible that transactions submitted via `submitpackage` was created somewhere else?

**Comment by instagibbs** [comment]
This PR is essentially focusing on the max fee check during submission to local mempool feature only. Wallet features to reduce accidental overpayment are also welcome, and may already exist in some forms?

File: src/validation.cpp [hunk 6]
```diff
@@ -464,6 +464,11 @@ class MemPoolAccept
          * policies such as mempool min fee and min relay fee.
          */
         const bool m_package_feerates;
+        /** Used for local submission of transactions to catch "absurd" fees
+         * due to fee miscalulation by wallets. std:nullopt implies unset, allowing any feerates.
+         * Any individual transaction failing this check causes immediate failure.
+         */
+        const std::optional<CFeeRate> m_max_sane_feerate;
```
Comment (glozow) [inline src/validation.cpp:471]: nit: I think calling this `m_client_maxfeerate` would make more sense. Don't think "sane" tells us very much, and it'd be good to specify that this comes from the client. Generally I don't think "sanity check" is the right term for these, just "check," as the user can provide a strange or no value and we don't judge whether it makes sense or not.

File: src/rpc/mempool.cpp [hunk 7]
```diff
@@ -828,6 +828,14 @@ static RPCHelpMan submitpackage()
                     {"rawtx", RPCArg::Type::STR_HEX, RPCArg::Optional::OMITTED, ""},
                 },
             },
+            {"maxfeerate", RPCArg::Type::AMOUNT, RPCArg::Default{FormatMoney(DEFAULT_MAX_RAW_TX_FEE_RATE.GetFeePerK())},
+             "Reject transactions whose fee rate is higher than the specified value, expressed in " + CURRENCY_UNIT +
+                 "/kvB.\nSet to 0 to accept any fee rate."},
+            {"maxburnamount", RPCArg::Type::AMOUNT, RPCArg::Default{FormatMoney(0)},
```
Comment (glozow) [inline src/rpc/mempool.cpp:834]: Probably good time to make a `DEFAULT_MAX_BURN_AMOUNT{0}`

File: src/validation.cpp [hunk 8]
```diff
@@ -1295,6 +1312,15 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::
             results.emplace(ws.m_ptx->GetWitnessHash(), MempoolAcceptResult::Failure(ws.m_state));
             return PackageMempoolAcceptResult(package_state, std::move(results));
         }
+
+        // Individual modified feerate exceeded caller-deemed "sane" max; abort
+        if (args.m_max_sane_feerate && CFeeRate(ws.m_modified_fees, ws.m_vsize) > args.m_max_sane_feerate.value()) {
```
Comment (glozow) [inline src/validation.cpp:1317]: I can see why we'd check individual feerate instead of chunk feerate, since we don't currently check whether there's a real CPFP happening and which transactions are in the chunk containing a CPFP. Could add a TODO item explaining this?

File: test/functional/rpc_packages.py [hunk 9]
```diff
@@ -356,5 +357,33 @@ def test_submitpackage(self):
...
+    def test_maxfeerate_maxburn_submitpackage(self):
+        node = self.nodes[0]
+        # clear mempool
+        deterministic_address = node.get_deterministic_priv_key().address
+        self.generatetoaddress(node, 1, deterministic_address)
+
+        self.log.info("Submitpackage maxfeerate arg testing")
+        chained_txns = self.wallet.create_self_transfer_chain(chain_length=2)
+        chain_hex = [t["hex"] for t in chained_txns]
+        pkg_result = node.submitpackage(chain_hex, 0.00000001)
```
Comment (glozow) [inline test/functional/rpc_packages.py:369]: Maybe instead of magic number, calculate/specify the parent feerate and make `maxfeerate` a little bit less?

File: test/functional/rpc_packages.py [hunk 10]
```diff
@@ -356,5 +357,33 @@ def test_submitpackage(self):
...
+        self.log.info("Submitpackage maxburnamount arg testing")
+        tx = tx_from_hex(chain_hex[1])
+        tx.vout[-1].scriptPubKey = b'a' * 10001 # scriptPubKey bigger than 10k IsUnspendable
+        chain_hex = [chain_hex[0], tx.serialize().hex()]
+        # burn test is run before any package evaluation; nothing makes it in and we get broader exception
+        assert_raises_rpc_error(-25, "Unspendable output exceeds maximum configured by user", node.submitpackage, chain_hex, 0, 0.1)
+        assert_equal(node.getrawmempool(), [])
+
+        # Relax the restrictions for both and send it; parent gets through as own subpackage
+        pkg_result = node.submitpackage(chain_hex, 50, 50)
```
Comment (glozow) [inline test/functional/rpc_packages.py:383]: Similarly, maybe specify `maxburnamount` as == the output amount? Imo a test should fail if I switch the check from `>` to `>=`.

File: src/rpc/mempool.cpp [hunk 11]
```diff
@@ -875,16 +896,24 @@ static RPCHelpMan submitpackage()
...
+                    }
+                }
+
                 txns.emplace_back(MakeTransactionRef(std::move(mtx)));
             }
             if (!IsChildWithParentsTree(txns)) {
                 throw JSONRPCTransactionError(TransactionError::INVALID_PACKAGE, "package topology disallowed. not child-with-parents or parents depend on each other.");
             }
 
+
```
Comment (glozow) [inline src/rpc/mempool.cpp:912]: random newline?

File: src/validation.cpp (hunk 6 above)
Comment (instagibbs) [reply src/validation.cpp]: done

File: src/rpc/mempool.cpp (hunk 7 above)
Comment (instagibbs) [reply src/rpc/mempool.cpp]: done

File: src/validation.cpp (hunk 8 above)
Comment (instagibbs) [reply src/validation.cpp]: added note

File: test/functional/rpc_packages.py (hunk 9 above)
Comment (instagibbs) [reply test/functional/rpc_packages.py]: done

File: test/functional/rpc_packages.py (hunk 10 above)
Comment (instagibbs) [reply test/functional/rpc_packages.py]: done

File: src/rpc/mempool.cpp (hunk 11 above)
Comment (instagibbs) [reply src/rpc/mempool.cpp]: removed

File: src/rpc/mempool.cpp [hunk 12]
```diff
@@ -828,6 +828,14 @@ static RPCHelpMan submitpackage()
                     {"rawtx", RPCArg::Type::STR_HEX, RPCArg::Optional::OMITTED, ""},
                 },
             },
+            {"maxfeerate", RPCArg::Type::AMOUNT, RPCArg::Default{FormatMoney(DEFAULT_MAX_RAW_TX_FEE_RATE.GetFeePerK())},
+             "Reject transactions whose fee rate is higher than the specified value, expressed in " + CURRENCY_UNIT +
+                 "/kvB.\nSet to 0 to accept any fee rate."},
+            {"maxburnamount", RPCArg::Type::AMOUNT, RPCArg::Default{FormatMoney(0)},
+             "Reject transactions with provably unspendable outputs (e.g. 'datacarrier' outputs that use the OP_RETURN opcode) greater than the specified value(default of 0), expressed in " + CURRENCY_UNIT + ".\n"
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply src/rpc/mempool.cpp]: done

File: src/rpc/mempool.cpp [hunk 13]
```diff
@@ -47,7 +47,7 @@ static RPCHelpMan sendrawtransaction()
              "Reject transactions whose fee rate is higher than the specified value, expressed in " + CURRENCY_UNIT +
                  "/kvB.\nSet to 0 to accept any fee rate."},
             {"maxburnamount", RPCArg::Type::AMOUNT, RPCArg::Default{FormatMoney(0)},
-             "Reject transactions with provably unspendable outputs (e.g. 'datacarrier' outputs that use the OP_RETURN opcode) greater than the specified value, expressed in " + CURRENCY_UNIT + ".\n"
+             "Reject transactions with provably unspendable outputs (e.g. 'datacarrier' outputs that use the OP_RETURN opcode) greater than the specified value(default of 0), expressed in " + CURRENCY_UNIT + ".\n"
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply src/rpc/mempool.cpp]: done

**Comment by ismaelsadeeq** [comment]
Code review ACK 83f34b4783565336341948affa42e3ec71242c4b

File: src/rpc/client.cpp (hunk 2 above)
Comment (ismaelsadeeq) [reply src/rpc/client.cpp:131]: Note  a good idea, because we have to restart whenever we want to update the threshold for an RPC call, can be resolved.

File: src/validation.cpp (hunk 1 above)
Comment (ismaelsadeeq) [reply src/validation.cpp]: I understand why we should not check the package fee rate here because, in some arrangements, e.g., a parent transaction P will be cfp'ed by some child E, and E can have another child F, which is on its own. 
Checking the package fee rate in this case is incorrect; we should instead evaluate using chunks fee rate later, as you said  👍🏾

File: src/rpc/mempool.cpp [hunk 14]
```diff
@@ -867,6 +876,19 @@ static RPCHelpMan submitpackage()
                                    "Array must contain between 1 and " + ToString(MAX_PACKAGE_COUNT) + " transactions.");
             }
 
+            // Fee check needs to be run with chainstate and package context
+            const CFeeRate max_raw_tx_fee_rate = request.params[1].isNull() ?
+                                                     DEFAULT_MAX_RAW_TX_FEE_RATE :
+                                                     CFeeRate(AmountFromValue(request.params[1]));
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply src/rpc/mempool.cpp:882]: to make sure I'm not misreading, this is a bump of the default feerate value by 10x? Making sure since I don't see any tests that had to be changed.

[the remark this answers is not in the document]
Comment (instagibbs) [reply src/rpc/mempool.cpp]: taken, thanks

**Comment by instagibbs** [comment]
rebased

**Comment by achow101** [comment]
ACK b6fe346c695d70b0d45d4a5e346aaa591cd2ef48

**Review by ismaelsadeeq — APPROVED** [review]
Re ACK https://github.com/bitcoin/bitcoin/commit/b6fe346c695d70b0d45d4a5e346aaa591cd2ef48 via [diff](https://github.com/bitcoin/bitcoin/compare/bf6fbea9794db5e8d0adce8ff6182574808cea54..b6fe346c695d70b0d45d4a5e346aaa591cd2ef48)

File: src/validation.cpp [hunk 15]
```diff
@@ -472,6 +472,11 @@ class MemPoolAccept
          * policies such as mempool min fee and min relay fee.
          */
         const bool m_package_feerates;
+        /** Used for local submission of transactions to catch "absurd" fees
+         * due to fee miscalulation by wallets. std:nullopt implies unset, allowing any feerates.
+         * Any individual transaction failing this check causes immediate failure.
```
Comment (murchandamus) [inline src/validation.cpp:477]: Should the `maxfeerate` perhaps rather apply to a transaction’s mining score in the package? That way, a child transactions bumping a large feerate could have a much larger nominal feerate, while staying in the sane range for the effective feerate.

**Review by murchandamus — COMMENTED** [review]
Concept ACK

File: test/functional/rpc_packages.py [hunk 16]
```diff
@@ -356,5 +357,34 @@ def test_submitpackage(self):
...
+        self.log.info("Submitpackage maxburnamount arg testing")
+        tx = tx_from_hex(chain_hex[1])
+        tx.vout[-1].scriptPubKey = b'a' * 10001 # scriptPubKey bigger than 10k IsUnspendable
+        chain_hex = [chain_hex[0], tx.serialize().hex()]
+        # burn test is run before any package evaluation; nothing makes it in and we get broader exception
+        assert_raises_rpc_error(-25, "Unspendable output exceeds maximum configured by user", node.submitpackage, chain_hex, 0, chained_txns[1]["new_utxo"]["value"] - Decimal("0.00000001"))
+        assert_equal(node.getrawmempool(), [])
+
+        # Relax the restrictions for both and send it; parent gets through as own subpackage
+        pkg_result = node.submitpackage(chain_hex, maxfeerate=0, maxburnamount=chained_txns[1]["new_utxo"]["value"])
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply test/functional/rpc_packages.py]: set it to `minrate_btc_kvb` which does the trick I think

File: test/functional/rpc_packages.py [hunk 17]
```diff
@@ -356,5 +357,34 @@ def test_submitpackage(self):
...
+        node = self.nodes[0]
+        # clear mempool
+        deterministic_address = node.get_deterministic_priv_key().address
+        self.generatetoaddress(node, 1, deterministic_address)
+
+        self.log.info("Submitpackage maxfeerate arg testing")
+        chained_txns = self.wallet.create_self_transfer_chain(chain_length=2)
+        minrate = min([chained_txn["fee"] / chained_txn["tx"].get_vsize() for chained_txn in chained_txns])
+        chain_hex = [t["hex"] for t in chained_txns]
+        pkg_result = node.submitpackage(chain_hex, maxfeerate=minrate-Decimal("0.00000001"))
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply test/functional/rpc_packages.py]: good catch, renamed variable to make it clearer, hopefully

File: src/validation.cpp [hunk 18]
```diff
@@ -472,6 +472,11 @@ class MemPoolAccept
          * policies such as mempool min fee and min relay fee.
          */
         const bool m_package_feerates;
+        /** Used for local submission of transactions to catch "absurd" fees
+         * due to fee miscalulation by wallets. std:nullopt implies unset, allowing any feerates.
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply src/validation.cpp]: done

File: src/validation.h [hunk 19]
```diff
@@ -274,13 +274,15 @@ MempoolAcceptResult AcceptToMemoryPool(Chainstate& active_chainstate, const CTra
...
+* @param[in]    test_accept         When true, run validation checks but don't submit to mempool.
+* @param[in]    max_sane_feerate    If exceeded by an individual transaction, rest of (sub)package evalution is aborted.
+*                                   Only for sanity checks against local submission of transactions.
 * @returns a PackageMempoolAcceptResult which includes a MempoolAcceptResult for each transaction.
 * If a transaction fails, validation will exit early and some results may be missing. It is also
 * possible for the package to be partially submitted.
 */
 PackageMempoolAcceptResult ProcessNewPackage(Chainstate& active_chainstate, CTxMemPool& pool,
-                                                   const Package& txns, bool test_accept)
+                                                   const Package& txns, bool test_accept, std::optional<CFeeRate> max_sane_feerate)
```
Comment (glozow) [inline src/validation.h:285]: nit: Seems like you are passing by reference in most places, which makes sense to me
```suggestion
                                                   const Package& txns, bool test_accept, const std::optional<CFeeRate>& max_sane_feerate)
```

**Review by glozow — COMMENTED** [review]
ACK 38f70ba6ac with some non-blocking nits

**Comment by instagibbs** [comment]
@glozow I'll address these if I touch the commits again

File: src/rpc/mempool.cpp [hunk 20]
```diff
@@ -823,6 +824,14 @@ static RPCHelpMan submitpackage()
                     {"rawtx", RPCArg::Type::STR_HEX, RPCArg::Optional::OMITTED, ""},
                 },
             },
+            {"maxfeerate", RPCArg::Type::AMOUNT, RPCArg::Default{FormatMoney(DEFAULT_MAX_RAW_TX_FEE_RATE.GetFeePerK())},
+             "Reject transactions whose fee rate is higher than the specified value, expressed in " + CURRENCY_UNIT +
+                 "/kvB.\nFee rates larger than 1BTC/kvB are rejected.\nSet to 0 to accept any fee rate."},
```
[the remark this answers is not in the document]
Comment (murchandamus) [reply src/rpc/mempool.cpp:829]: Opened in #29661

**Review by murchandamus — COMMENTED** [review]
LGTM, code review ACK 38f70ba6ac86fb96c60571d2e1f316315c1c73cc

**Comment by ismaelsadeeq** [comment]
Re-ACK https://github.com/bitcoin/bitcoin/commit/38f70ba6ac86fb96c60571d2e1f316315c1c73cc 👍🏾

**Comment by instagibbs** [comment]
follow-ups for this here https://github.com/bitcoin/bitcoin/pull/29722 @glozow

File: src/validation.cpp [hunk 21]
```diff
@@ -472,6 +472,11 @@ class MemPoolAccept
...
+         * Any individual transaction failing this check causes immediate failure.
+         */
+        const std::optional<CFeeRate> m_client_maxfeerate;
```
[the remark this answers is not in the document]
Comment (stickies-v) [reply src/validation.cpp:479]: I think the [current](https://github.com/bitcoin/bitcoin/blob/2e1c84b333dc744dec04b783dc77b24a55e528d8/src/validation.cpp#L449) memory layout requires 56 bytes for an `ATMPArgs` object:

```cpp

    struct ATMPArgs {
        const CChainParams& m_chainparams;                      // 8 bytes
        const int64_t m_accept_time;                            // 8 bytes
        const bool m_bypass_limits;                             // 1 bytes
                                                                // 7 bytes padding 
        std::vector<COutPoint>& m_coins_to_uncache;             // 8 bytes
        const bool m_test_accept;                               // 1 bytes
        const bool m_allow_replacement;                         // 1 bytes
        const bool m_package_submission;                        // 1 bytes
        const bool m_package_feerates;                          // 1 bytes
                                                                // 4 bytes padding
        const std::optional<CFeeRate> m_client_maxfeerate;      // 9 bytes
                                                                // 7 bytes padding
    }; // TOTAL: 7 * 8 = 56 bytes
```

With optimal (?) member ordering, we could get this down to 48 bytes, I think:

```cpp
    struct ATMPArgs {
        const CChainParams& m_chainparams;                      // 8 bytes
        const int64_t m_accept_time;                            // 8 bytes
        std::vector<COutPoint>& m_coins_to_uncache;             // 8 bytes
        const std::optional<CFeeRate> m_client_maxfeerate;      // 9 bytes
                                                                // 7 bytes padding
        const bool m_bypass_limits;                             // 1 bytes
        const bool m_test_accept;                               // 1 bytes
        const bool m_allow_replacement;                         // 1 bytes
        const bool m_package_submission;                        // 1 bytes
        const bool m_package_feerates;                          // 1 bytes
                                                                // 3 bytes padding
    }; // TOTAL: 6 * 8 bytes = 48 bytes
```

However, given that afaik we generally only spawn one `ATMPArgs` at the same time, I don't think we should change the member ordering just for memory footprint here, and it's fine to leave as is? Likewise, adding actual struct packing (e.g. with #pragma) almost certainly doesn't seem worth the performance trade-off (didn't run any benchmarks).

[the remark this answers is not in the document]
Comment (glozow) [reply src/validation.cpp:479]: Nice, thanks @stickies-v! Agree it's not really worth it when we only ever have 1 at a time.
